### PR TITLE
Added likes and connected users, posts, and likes

### PR DESCRIPTION
**Pull request by ryang72**
## Changes
- added like model and controller (need to do something similar for comments)
- connected posts and users
- improved ui, only users can create posts / like

Like implementation is from this guide: https://medium.com/full-taxx/how-to-add-likes-to-posts-in-rails-e81430101bc2
seems pretty simple, can replace if doesn't scale enough

okay honestly for likes id rather just use this gem: https://github.com/ryanto/acts_as_votable
and not really think about this rather janky implementation from above

So basically we really need to get more interactions before we do load testing. Even if the code is messy. I prob missed a lot of loose ends that need to be cleaned up but at least we have a working baseline to Tsung on top of.

#### Attached issue?
#23 #27 #28

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 class PostsController < ApplicationController
+  before_action :find_user, only: [:create]
```
Comment (zhaoalex) [inline app/controllers/posts_controller.rb]: Don't need these `before_action`s if you only use them for one method, just go ahead and put them in the method

File: app/models/user.rb [hunk 2]
```diff
@@ -22,4 +22,9 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+
+  # user has many posts, once account is deleted all posts made should
```
Comment (zhaoalex) [inline app/models/user.rb]: why? shouldn't we keep posts even if the account is deleted?

File: app/views/posts/new.html.erb [hunk 3]
```diff
@@ -13,3 +16,7 @@
     <%= form.submit %>
   </p>
 <% end %>
+<% end %>
+<% if not user_signed_in? %>
```
Comment (zhaoalex) [inline app/views/posts/new.html.erb]: `else` > `if not`

File: app/views/posts/new.html.erb [hunk 4]
```diff
@@ -13,3 +16,7 @@
...
+  <%= link_to 'Login in to make a post!', '/users/login' %>
```
Comment (zhaoalex) [inline app/views/posts/new.html.erb]: Login, not "login in"

File: app/views/posts/new.html.erb [hunk 5]
```diff
@@ -1,4 +1,7 @@
-<%= form_with scope: :post, url: posts_path, local: true do |form| %>
+<% if user_signed_in? %>
+  <div>Signed in as... <%= current_user.email %></div>
+  <%= form_with scope: :post, url: posts_path, local: true do |form| %>
+  <%= form.hidden_field :user_id, value: current_user.id %>
```
Comment (zhaoalex) [inline app/views/posts/new.html.erb]: Please fix indentation here

File: app/models/post.rb [hunk 6]
```diff
@@ -5,16 +5,19 @@
 #  id         :integer          not null, primary key
 #  author     :string           not null
```
Comment (zhaoalex) [inline app/models/post.rb]: not something you need to do in this PR, but we should remove the author field cuz user_id takes care of that

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -8,16 +10,28 @@ def show
   end
 
   def create
-    @post = Post.new(post_params) # TODO: author, location
-    @post.likes = 0
-    @post.author = 'garbo'
+    @post = current_user.post.new(post_params) # TODO: author, location
+    @post.author = current_user.email
```
Comment (zhaoalex) [inline app/controllers/posts_controller.rb]: again, not something you need to do in this PR, but we should prob remove the author field cuz user_id takes care of that

File: app/controllers/posts_controller.rb [hunk 8]
```diff
@@ -8,16 +10,28 @@ def show
...
+  end
+
   private
 
+  def find_post
+    @post = Post.find(params[:id])
+  end
+
+  def find_user
+    @user = User.find_by_email(params[current_user.email])
```
Comment (zhaoalex) [inline app/controllers/posts_controller.rb]: haven't looked into devise much but can't we just say `@user = current_user` here?

File: app/views/posts/show.html.erb [hunk 9]
```diff
@@ -1,3 +1,18 @@
...
+  <div>Signed in as... <%= current_user.email %></div>
+<%end%>
 <h1><%= @post.title %></h1>
-<h2><%= @post.author %></h2>
+<h3><%= @post.author %></h3>
+<h4>spice level: <%= @post.likes.count %></h4>
+<% if user_signed_in? %>
+  <% pre_like = @post.likes.find { |like| like.user_id == current_user.id} %>
+  <% if pre_like %>
+    <%= button_to 'dilute', post_like_path(@post, pre_like), method: :delete %>
```
Comment (zhaoalex) [inline app/views/posts/show.html.erb]: is this a separate button to unlike? in the future let's just click the same button again to unlike

File: app/views/posts/show.html.erb [hunk 10]
```diff
@@ -1,3 +1,18 @@
+<% if user_signed_in? %>
+  <div>Signed in as... <%= current_user.email %></div>
+<%end%>
 <h1><%= @post.title %></h1>
-<h2><%= @post.author %></h2>
+<h3><%= @post.author %></h3>
+<h4>spice level: <%= @post.likes.count %></h4>
+<% if user_signed_in? %>
+  <% pre_like = @post.likes.find { |like| like.user_id == current_user.id} %>
```
Comment (zhaoalex) [inline app/views/posts/show.html.erb]: prefer `find_by(user_id: current_user.id)`

File: app/views/posts/show.html.erb [hunk 11]
```diff
@@ -1,3 +1,18 @@
...
+  <% if pre_like %>
+    <%= button_to 'dilute', post_like_path(@post, pre_like), method: :delete %>
+  <% else %>
+    <%= button_to 'enrich', post_likes_path(@post), method: :post %>
+  <% end %>
+<% end %>
 <p><%= @post.body %></p>
+<% if current_user.email == @post.author %>
```
Comment (zhaoalex) [inline app/views/posts/show.html.erb]: should check on user id and not author name

File: db/migrate/20191031074056_create_likes.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+class CreateLikes < ActiveRecord::Migration[6.0]
+  def change
+    create_table :likes do |t|
+      t.references :post, null: false, foreign_key: true
```
Comment (zhaoalex) [inline db/migrate/20191031074056_create_likes.rb]: rails prefers handling FKs in application layer, not DB layer; just a heads up

File: db/migrate/20191031084518_add_user_id_to_post.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+class AddUserIdToPost < ActiveRecord::Migration[6.0]
+  def change
+    add_reference :posts, :user, null: false, foreign_key: true
```
Comment (zhaoalex) [inline db/migrate/20191031084518_add_user_id_to_post.rb]: same for FK here

File: spec/factories/likes.rb [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+#
+# Indexes
+#
+#  index_likes_on_post_id  (post_id)
+#  index_likes_on_user_id  (user_id)
+#
+
+FactoryBot.define do
+  factory :like do
+    post { nil }
```
Comment (zhaoalex) [inline spec/factories/likes.rb]: Already set up a factory for post, so you can just create(post) here

Comment (zhaoalex) [reply spec/factories/likes.rb]: not too important tho

File: app/models/user.rb [hunk 15]
```diff
@@ -22,4 +22,9 @@ class User < ApplicationRecord
...
+  # be deleted as well
+  has_many :post, dependent: :destroy
```
Comment (zhaoalex) [inline app/models/user.rb]: `:posts`

File: app/views/posts/new.html.erb (hunk 3 above)
Comment (ryang72) [reply app/views/posts/new.html.erb]: I tried using `else` but it would error? Specifically I tried `<% if ...%> <% else %> <% end %>` which is what someone else did online

File: db/migrate/20191031084518_add_user_id_to_post.rb (hunk 13 above)
Comment (zhaoalex) [reply db/migrate/20191031084518_add_user_id_to_post.rb]: not too important tho

File: db/migrate/20191031074056_create_likes.rb (hunk 12 above)
Comment (zhaoalex) [reply db/migrate/20191031074056_create_likes.rb]: not too important tho
